### ISSUE TITLE
Clarification on context.Context with radix.Cluster

### ISSUE THREAD

**Issue by cep21** [mediocregopher/radix#129]
I read https://github.com/mediocregopher/radix.v2/issues/67#issuecomment-321370502 and understand you do not want to support context in your library, but I would like to see how I can best support it in my own use of radix.

You added https://github.com/mediocregopher/radix.v2/commit/ae7309086d191442b36bf69f7f5eeca5fdbd329e but it appears those are on a per client basis.  I have many threads using the cluster client at once and each request would need its own timeout, rather than adjusting the timeout of the cluster multiple times.

I understand it's difficult to fully support context.Done(), since it's not part of the `io.Write` API, but Deadline alone could be supported (this is how most TCP libraries support context).  (Context is supported on the TCP dial natively https://golang.org/pkg/net/#Dialer.DialContext)

If the additional function `func (c *Cluster) DoCtx(a Action) error {` is off the table, then maybe we could brainstorm how I could support context outside of changing `mediocregopher/radix`

I imagine an API to support context myself would be to change
```go
c.Do(Cmd(&vgot, "GET", k))
```

Into 
```go
c.Do(CmdCtx(ctx, &vgot, "GET", k))
```

Then I just need to correctly implement CmdCtx.  I think that works with

```go
// Action can perform one or more tasks using a Conn
type CtxAction struct {
	Wraps CmdAction
	Ctx context.Context
}

func (cx *CtxAction) Keys() []string {
	return cx.Wraps.Keys()
}

func (cx *CtxAction) Run(c Conn) error {
	deadline, ok := cx.Ctx.Deadline()
	if ok {
		if err := c.NetConn().SetDeadline(deadline); err != nil {
			return err
		}
	}
	return c.Do(cx)
}

func (cx *CtxAction) MarshalRESP(w io.Writer) error {
	return cx.Wraps.MarshalRESP(w)
}

func (cx *CtxAction) UnmarshalRESP(br *bufio.Reader) error {
	return cx.Wraps.UnmarshalRESP(br)
}

var _ CmdAction = &CtxAction{}
```

My only concern here is what guarantees does the new radix library have about the order of operations between `Run` -> `UnmarshalRESP`

After this, the next part is hooking this up to a tracing system such as AWS x-ray or Jaeger, or Datadog's APM.  Both of these libraries require access to context.  I could get this access if your internal tracing exposed the 'Action' object that triggered operations.

I notice you have `ClusterTrace` and `PoolTrace`.  I could get the context if the Action object was given to these.  The action could be nil to indicate a trigger not from an action.  But for example if `DoCompleted func(PoolDoCompleted)` included the `Action` I could do a type cast on action to get an optional context

```go
func (cx *CtxAction) Context() context.Context {
    return ctx.Ctx
}

type HasContext interface {
  Context() context.Context
}

DoCompleted =  func(p PoolDoCompleted) {
  if withCtx, ok := p.Action.(HasContext); ok {
    jaeger.Trigger(withCtx.Context(), blah,blah)
  }
}
```

How are you envisioning the best way someone could support context with radix, now that the library has changed to be more interface driven?

**Comment by mediocregopher** [comment]
> I understand it's difficult to fully support context.Done(), since it's not part of the io.Write API, but Deadline alone could be supported (this is how most TCP libraries support context).

That's pretty much the crux of the issue. It's straightforward enough to support Deadline, but supporting cancellation is not possible (as far as I know), and I didn't want to put only partial support for Contexts as I felt that would cause more confusion than anything.

> Then I just need to correctly implement CmdCtx

Yup! Your implementation looks about right, except for your `Run` method, which I think causes an infinite loop right now (`Conn.Do` merely calls `Action.Run(conn)`). I would change it to:

```go
	if deadline, ok := cx.Ctx.Deadline(); ok {
		if err := c.NetConn().SetDeadline(deadline); err != nil {
			return err
		}
	}
        if err := c.Encode(cx); err != nil {
                return err
        }
        return c.Decode(cx)
```

I would also note that, if `CtxAction` was used in a `Pipeline` it would effectively bypass the deadline handling. This is because pipelining only uses `MarshalRESP` and `UnmarshalRESP`. And since radix will implicitly pipeline `CmdAction`s in the default pool implementation, given default options, your Ctx would generally not get used. You've got two options there:

1) Disable implicit pipelining with [PoolPipelineWindow(0)](https://godoc.org/github.com/mediocregopher/radix#PoolPipelineWindow).

2) Don't implement `MarshalRESP` and `UnmarshalRESP` in your `CtxAction`, which will mean it only implements `Action`, and therefore will never be eligible for pipelining.

It's up to how you plan on using it which one you choose.

> My only concern here is what guarantees does the new radix library have about the order of operations between Run -> UnmarshalRESP

I think the above answers that. UnmarshalRESP will get called within `Conn.Decode`, which gets called within `Run`. If pipelining is being used then `Run` won't be called at all, only `UnmarshalRESP`.

>  I could get this access if your internal tracing exposed the 'Action' object that triggered operations.

I think that's a good idea. [ClusterRedirected](https://godoc.org/github.com/mediocregopher/radix/trace#ClusterRedirected) and [PoolDoCompleted](https://godoc.org/github.com/mediocregopher/radix/trace#PoolDoCompleted) would both benefit greatly by having the `Action` as a field. If you wanted to submit a PR for that I'd be happy to look at it :)

> How are you envisioning the best way someone could support context with radix, now that the library has changed to be more interface driven?

Unfortunately it's still somewhat difficult, but I think you've figured out most of it. It will require giving some features up, like implicit pipelining, which is unfortunate. But honestly even if I was building this from the ground up I'm not sure if I _could_ support implicit pipelining with Contexts. Which Context would I use for any given pipeline? Pick the one with the soonest deadline? That seems wrong to me. But the nice thing about radix is that you could drop in your own Pool implementation if you wanted, one which treated contexts as you liked, and everything else would still work.

And even without implicit pipelining radix is pretty darn fast, so you're not losing a whole lot.

Hopefully that cleared things up, definitely let me know if you have any other questions, and sorry if I'm a bit slow to answer.

**Comment by cep21** [comment]
Maybe I'm over thinking this.  Just a few more questions.  Thanks for the time!

> But the nice thing about radix is that you could drop in your own Pool implementation if you wanted, one which treated contexts as you liked, and everything else would still work.

I like your pool code and would love to integrate with it somehow rather than fork it :)



I noticed parts of the code where you break out of using the interface itself to implementations like this: https://github.com/mediocregopher/radix/blob/1594fa6fbca09c298ed1379f6a6a3b39fba45fd1/pipeliner.go#L85

Can I change these to optional interface checks instead?  Are there other places in the code where you check against explicit implementations?

```
        type CommandName interface {
              CmdName() string
        }
	if cmdA, ok := a.(CommandName); ok {
		return !blockingCmds[strings.ToUpper(cmdA.CmdName())]
	}
```

My concern with overwriting the interfaces is that I may have something that works with how the interfaces are used today, but may not later if you refactor the order of operations.  Is there a contract on the interfaces that guarantees the order things will work on.  Something like "If you pass an action that works like this, I promise it will always work like _this_".  If so, it would be good to document.

For example, I tried to implement CmdCtx but pipeline didn't obey what I thought the contract of `Run` was.  If `mediocregopher/radix` didn't have the pipeline concept, my code would have worked as expected, until you wrote pipeline which would have broke the assumptions (guarantees (?)) of the interface I had made and broke CmdCtx. (or maybe I missed this part)

> But honestly even if I was building this from the ground up I'm not sure if I could support implicit pipelining with Contexts. Which Context would I use for any given pipeline?

That's a really good point.  The safest would be to use the longest deadline of the bunch.  Pipeline could have a function that coalesces contexts (with default of returning longest deadline) and people could change that to pick how they want that to behave.


If I added

```
type ActionWithContext interface {
  Action
  Context() context.Context
}
```

This would be backwards compatible.  It would give extenders of `radix` a base to work with for supporting context.

**Comment by mediocregopher** [comment]
> Can I change these to optional interface checks instead? Are there other places in the code where you check against explicit implementations?

Not off the top of my head, I tried to keep them to a minimum. We decided it was ok for now with pipeliner for a few reasons, one of which is that, as you allude to, it could mess with custom implementations of Action or CmdAction. I'll think about if it's worth the API space to have a way to signal to Pool that a CmdAction can be pipelined (it would be something like `CmdActionBlocks` though, not based on the name). But every new public type is another thing which might break when something is changed later, so I'm always hesitant to add them.

> Something like "If you pass an action that works like this, I promise it will always work like this". If so, it would be good to document.

It seems that, in this case, the issue is that the docs don't really give an idea of how all the pieces fit together, specifically regarding the dichotemy between `Run` and `MarshalRESP`/`UnmarshalRESP` and the different use-cases they support. I think if they did a better job of that it would be clear what the contract is, because honestly it couldn't be refactored to be any other way.

I'll make an issue to remind me to re-look at the docs in this area.

> For example, I tried to implement CmdCtx but pipeline didn't obey what I thought the contract of Run was. If mediocregopher/radix didn't have the pipeline concept, my code would have worked as expected

I think you misunderstood. The code you wrote was broke no matter what, it would have infinite looped with or without implicit pipelining. I think that was an issue of poor documentation on my part and not a break in backwards compatibility.

The code I posted would work (hypothetically, I never actually tested it) with or without implicit pipelining. I actually misspoke when I said that implicit pipelining would cause the default pool to generally ignore your deadline, I had forgotten that (as you pointed out) it explicitly checks for a `*cmdAction`. So it would have worked correctly, but would not have benefited from implicit pipelining.

> That's a really good point. The safest would be to use the longest deadline of the bunch. Pipeline could have a function that coalesces contexts (with default of returning longest deadline) and people could change that to pick how they want that to behave.

If you're talking about the explicit [Pipeline](https://godoc.org/github.com/mediocregopher/radix#Pipeline) then I think it'd be best to just implement that on your end for now, it seems like something which will have far more edge-cases than can be predicted ahead of time. And Pipeline is fairly simple.

> Maybe I'm over thinking this. Just a few more questions. Thanks for the time!

It's all good, sorry the docs are butt :)

**Comment by cdepauw** [comment]
A bit late to the party, but is CLIENT UNBLOCK an option here? Cfr. https://redis.io/commands/client-unblock

For now I'm using a custom CmdContext that basically wraps around a CmdAction and closes the connection if the context were to cancel before the command is complete. It doesn't work with pipelining though and is a bit harsh perhaps, but fits my use case for blocking commands and long polling Redis.

**Comment by mediocregopher** [comment]
> A bit late to the party, but is CLIENT UNBLOCK an option here? Cfr. https://redis.io/commands/client-unblock

I don't think that's a workable solution. For one it would require managing all the client ids of the connections in the pool, which is currently not done and would add a non-trivial amount of overhead. Also it doesn't really guarantee that the connection will actually become unblocked (e.g. if there's some network error that causes the server to drop the connection, but the client still thinks it's open), so we'd still be left with the same problem.

> For now I'm using a custom CmdContext that basically wraps around a CmdAction and closes the connection if the context were to cancel before the command is complete.

That actually sounds pretty reasonable, in the normal case if a connection were to timeout then Pool would also close it, and I think most other db drivers behave the same. I think you could even support pipelining by having the go-routine be launched in the `CmdContext` function, storing `done` on the struct, and closing it after _either_ `Run` or `UnmarshalRESP` has been called. (I think the docs specify somewhere that a `CmdAction` shouldn't be used again after `UnmarshalRESP` has been called, so there shouldn't be any conflicts there, I'll have to double check though).
